fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/metadata/Index.java

```diff
@@ -431,6 +433,32 @@ public void trimPrimaryKey(List<?> primaryKeys) {
         }
     }
 
+
+    /**
+     * Compute the full key of an index (given that the index is a value index).
+     *
+     * @param primaryKey primary key of the records the index ranges over. The primary key is used to determine
+     *        parts in the index definition that already contain parts of the primary key. All primary key components
+     *        that are not already part of the index key are appended to the index key.
+     * @return a {@link KeyExpression} describing the <em>full</em> index key as stored
+     * TODO the implementation is currently restricted to only work on value indexes. We need to make this generally
+     * TODO applicable to all index types in appropriate ways.
+     */
+    @Nonnull
+    public KeyExpression fullKey(@Nullable final KeyExpression primaryKey) {
+        Preconditions.checkArgument(getType().equals(IndexTypes.VALUE));
```

Comment:
Really this wants to somehow check `PlannableIndexTypes.getValueTypes`, to allow for extended value types.

This method ought to do like https://github.com/FoundationDB/fdb-record-layer/blob/master/fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/RecordQueryPlanner.java#L440-L442 and take just the key part of key-with-value, which isn't _really_ a key expression, but does appear in the root expression.

Reply:
This is now handled in the key expression expansion visitors.